### DISCUSSION TITLE
How to group and Take within a single query? [Conversations -> Messages]

### DISCUSSION THREAD

**Discussion by asusralis** [ravendb/ravendb#12697] [Q&A] [answered]
class User
    {
        public string Name { get; set; }
        public List<string> ConversationIds { get; set; }
    }

    class Message
    {
        public string Id { get; set; }
        public string ConversationId { get; set; }
        public DateTime Created { get; set; }
    }

    class Conversation
    {
        public string Id { get; set; }
    }

Would it be possible to get the x most recent messages, grouped, out of a collection of conversations? So for example, if ConversationIds contained Conversation-1-A and Conversation-2-A, it would return up to 10 of the most recent messages out of each conversations; a maximum of 20 messages grouped within their ConversationId.

I have seen the "How to preform group by query" document page, but I am not sure if it's possible to do in one query. I tried to make a conversation in google groups, but I'm not sure if it was posted... I did not see it. Sorry if this is a duplicate, but if it is I think I'd rather use this instead. Either way, thank you for any help!

**Comment by ayende** [comment, marked as answer]
You can do it, but it takes a couple of steps. The easiest way to handle that is to use a Map/Reduce index to gather all the messages for the conversation. 

```
// map 
from m in docs.Messages
select new {
 Messages= new[]{ new { m.Id, m.Created} },
 m.ConversationId
}

// reduce 
from r in results
group r by r.ConversationId
into g
select new 
{
   ConversationId = g.Key,
   Messages = g.SelectMany(x=>x.Messages).OrderByDescending(x=>x.Created).Take(15)
}
```

What this will do is group the messages per conversation and give you the 15 most recent ones.

You can the query it using:

```
from index 'RecentMessagesByConversationId'
where ConversationId in ($id1, $id2)
include Messages[].Id
```

That gives you the recent messages from both conversations quickly and in one shot to the database.

**Reply by asusralis** [reply]
Thank you for the help! This is what I was looking for. Sorry if this is a silly question, but how scalable is this method of getting the most recent messages? Would this need to query all existing messages or would it only have to query messages inside the group? I do not know much about databases, so I'm not sure if this would be the preferably way to query messages or if there is another, better way of doing so.

**Reply by arekpalinski** [reply]
I believe the answer to your question about scalability is to understand how actually map-reduce indexes work under the covers. You can read about it in the following resources (I hope you will find it interesting):

- https://ayende.com/blog/174051/the-design-of-ravendb-4-0-you-cant-see-the-map-reduce-from-all-the-trees
- https://ayende.com/blog/179938/ravendb-4-0-unsung-heroes-map-reduce
- https://ravendb.net/learn/inside-ravendb-book/reader/4.0/11-mapreduce-and-aggregations-in-ravendb

In general, the idea behind map-reduce indexes is to rebuild only relevant and small set of intermediate group by results in order to update the final aggregation. Also note that during the query we do the search using the already computed and indexed aggregation result.

Also as written in our docs (https://ravendb.net/docs/article-page/5.2/csharp/indexes/map-reduce-indexes):

"The idea behind map-reduce indexing is that aggregation queries using such indexes are very cheap. The aggregation is performed only once and the results are stored inside the index. Once new data comes into the database or existing documents are modified, the map-reduce index will keep the aggregation results up-to-date. The aggregations are never done during querying to avoid expensive calculations that could result in severe performance degradation. When you make the query, RavenDB immediately returns the matching results directly from the index."

**Reply by asusralis** [reply]
Thank you for all the information! Wow, this is such a cool feature - it was exactly what I was looking for. And not only that, I also know other queries that could use this as well. 

Hopefully I didn't miss this in what you sent me, but - would this make the database more memory bound and less cpu bound? I would assume mapping takes memory but trades not having to iterate entire collections when doing a query. This is typically preferable when you can have this tradeoff, yes? Thank you again for all of this.

**Reply by arekpalinski** [reply]
In general, the indexing in RavenDB is built this way to shift the cost of computation (CPU) to the indexing (where you can do it just once) instead of calculating query results each time when a query is run. I believe this chapter will give you all necessary details to understand our indexing / querying approach: https://ravendb.net/learn/inside-ravendb-book/reader/4.0/10-static-indexes-and-other-advanced-options

**Reply by asusralis** [reply]
This makes sense about shifting cost of computation. Now that I know about this, I have one more question: is it possible to, say, pass a parameter into the reduce part? I need to do paging on specific conversations, but right now I'm doing a query on all messages. So something from:

    Messages = g.SelectMany(x => x.Messages).OrderByDescending(x => x.Created).Take(10)

to

    Messages = g.SelectMany(x => x.Messages).OrderByDescending(x => x.Created).Skip(10 * page).Take(10)

this would allow me to get the initial messages and do paging all on the same index and avoid having to query through all messages each time.

**Reply by asusralis** [reply]
Oh, well, is this actually the best way to do it? It seems to work - after removing the Take from the reduce, of course. 

            var conversations = await session
                ...
                .Select(x => new
                {
                    ConversationId = x.ConversationId,
                    Messages = x.Messages.Skip(1).Take(3).ToList()
                })
                .ToListAsync();

**Reply by ayende** [reply]
No, that isn't a good idea. At least if you have unbounded conversations.

What it does, it applies the filter at the query time. But the reduce needs to deal with ALL the messages in the conversation.

If you have a known upper bound, that is better. Otherwise, you might want to use this to show the recent messages in a conversation, and to go deeper, use a separate query.

**Reply by asusralis** [reply]
Darn! If the messages are already mapped and ordered, why would skipping and taking from this list be bad? Would it ever be possible to support this in the future to avoid querying such a large collection?

**Reply by ayende** [reply]
There isn't a major cost to querying a large collection, given that you'll be using an index.
I assume that given your UI, you'll need:

* Show the recent conversations with the most recent messages (the map/reduce handles that).
* Show the messages for a particular conversation, which is:

```
from Messages
where ConversationId = $conId and Date between $start and $end
order by Date desc
```
